### PR TITLE
model: General Class for DecoderOnlyModel

### PR DESCRIPTION
**Pull request by rebel-thkim**
# Pull Request Description

## Type of Change

- [x] New Model Support
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Performance improvement
- [x] Code refactoring
- [ ] Other (please describe):

## Checklist

- [x] I have performed a self-review of my own code
- [x] I have added tests that prove my fix is effective or that my feature works (If needed)

## Related Issues

----
# Conventional commit
```
type(optional scope): description
```
----
# Type candidate
  - Model Updates
    - `model`: Adding New models or Bugfix for existing models
      - ex) Add LlavaNext 
      - ex) Bugfix Whisper
  - Enhancements
    - `performance`: Optimizing some models or this library itself
      - ex) Loading RBLNModel faster
      - ex) Optimizing Memory Usage of DecoderOnlyModel
  - Code Refactor
    - `refactor`: Re-arrange class architecture, or more.
      - ex) Refactor Seq2Seq
  - Documentation
    - `doc`: Update docstring only     
  - Library Dependencies
    - `dependency`: Update requirements, something like that.
  - Other
    - `other`: None of above.
      - ex) ci update
      - ex) pdm update

### CODE REVIEW

File: src/optimum/rbln/transformers/models/decoderonly/modeling_decoderonly.py [hunk 1]
```diff
@@ -1085,23 +815,520 @@ def _create_runtimes(
...
+        save_dir_path: Path,
+        subfolder: str,
+        rbln_config: RBLNDecoderOnlyModelForCausalLMConfig,
+    ):
+        # If you are unavoidably running on a CPU rather than an RBLN device,
+        # store the torch tensor, weight, etc. in this function.
+        if rbln_config.use_inputs_embeds:
+            save_dict = {}
+            save_dict["embed_tokens"] = model.get_input_embeddings().state_dict()
+            torch.save(save_dict, save_dir_path / subfolder / "torch_artifacts.pth")
```
Comment (rebel-jongho) [inline src/optimum/rbln/transformers/models/decoderonly/modeling_decoderonly.py:1020]: Duplicate of its parent class

File: src/optimum/rbln/transformers/models/decoderonly/modeling_decoderonly.py [hunk 2]
```diff
@@ -1085,23 +815,520 @@ def _create_runtimes(
...
+                if quantization:
+                    quantization.maybe_reset_quantization_env()
+
+        wrapped_model.phase = "prefill"
+        compiled_prefill = compile_model(
+            wrapped_model, prefill_compile_config, prefill_example_inputs, context, rbln_config.quantization
+        )
+        compiled_models = {"prefill": compiled_prefill}
+
+        if "decode" in rbln_config.phases:
```
Comment (rebel-jongho) [inline src/optimum/rbln/transformers/models/decoderonly/modeling_decoderonly.py:1158]: Are there any cases that `"decode"` `NOT in` `rbln_config.phases`?

File: src/optimum/rbln/transformers/models/decoderonly/configuration_decoderonly.py [hunk 3]
```diff
@@ -186,38 +177,197 @@ def __init__(
...
+            use_position_ids,
+            attn_impl,
+            kvcache_partition_len,
+            kvcache_block_size,
+            prefill_chunk_size,
+            kvcache_num_blocks,
+            cache_impl,
+            sliding_window,
+            sliding_window_layers,
+            **kwargs,
```
Comment (rebel-jongho) [inline src/optimum/rbln/transformers/models/decoderonly/configuration_decoderonly.py:342]: use key=value

File: src/optimum/rbln/transformers/models/decoderonly/configuration_decoderonly.py [hunk 4]
```diff
@@ -186,38 +177,197 @@ def __init__(
...
+        )
+
+        self.quantization = quantization or {}
+        if self.quantization and isinstance(self.quantization, dict):
+            self.quantization = RBLNQuantizationConfig(**self.quantization, model_config=self)
+
+        self.phases = phases or ["prefill", "decode"]
+
+        if "decode" not in self.phases:
```
Comment (rebel-jongho) [inline src/optimum/rbln/transformers/models/decoderonly/configuration_decoderonly.py]:
```suggestion
        if "decode" in self.phases:
```
??

File: src/optimum/rbln/transformers/models/decoderonly/modeling_decoderonly.py [hunk 5]
```diff
@@ -888,14 +656,12 @@ def get_input_info(
...
-            )
+            input_info.extend([("block_tables", [max_block_cnt], "int16")])
         if rbln_config.cache_impl in ["hybrid", "sliding_window"]:
-            input_info.extend([("local_block_tables", [1] if is_prefill else [batch_size, 1], "int16")])
+            input_info.extend([("local_block_tables", [1], "int16")])
 
-        # 4. query_position
-        if is_prefill:
+        # 4. query_position for sliding window attention
+        if rbln_config.cache_impl in ["sliding_window", "hybrid"]:
```
[the remark this answers is not in the document]
Comment (rebel-jongho) [reply src/optimum/rbln/transformers/models/decoderonly/modeling_decoderonly.py:669]: is there any `Model` with no swa?

File: src/optimum/rbln/transformers/models/qwen3/modeling_qwen3.py [hunk 6]
```diff
@@ -63,315 +54,6 @@ def forward(self, *args, **kwargs):
         return super().forward(*args, **kwargs)
```
Comment (rebel-jongho) [inline src/optimum/rbln/transformers/models/qwen3/modeling_qwen3.py:54]: @rebel-kblee Just curious, why the above line overwrites `return_dict`?

Comment (rebel-kblee) [reply src/optimum/rbln/transformers/models/qwen3/modeling_qwen3.py:54]: In Qwen3ForCausalLM has no return_dict params, so it always return with dict (return_dict=True). But our RBLNDecoderOnlyModelForCausalLM forward() uses return_dict

**Comment by rebel-kblee** [comment]
> One major modification is required.
> 
> Currently, `DecoderOnlyWrapper` handles both `DecoderOnlyModel` and `DecoderOnlyForCausalLM` using the is_causal_lm flag. However, this is conceptually incorrect. Please split `DecoderOnlyWrapper` into `DecoderOnlyModelWrapper` and `DecoderOnlyForCausalLMWrapper` so that `Model` can properly be part of `CausalLM`.

We had the same thought. But, if we had proceeded with that approach, there would have been too much repetition of duplicate code, so we decided to maintain the current state and refactor. Do you think that approach is acceptable?

File: src/optimum/rbln/transformers/models/decoderonly/modeling_decoderonly.py [hunk 7]
```diff
@@ -447,94 +449,66 @@ def prefill_forward(
...
-                dec_attn_mask=dec_attn_mask,
-                block_tables=block_tables,
-                free_block_pool=free_block_pool,
-                rbln_config=self.rbln_config,
+        # TODO: add prefill runtime class.
+        self.prefill_decoder = RBLNPytorchRuntime(runtime=self.model[0])
+
+        # attributes for prefill
+        self.use_global_attention = self.rbln_config.cache_impl in ["static", "hybrid"]
+        self.use_local_attention = self.rbln_config.cache_impl in ["sliding_window", "hybrid"]
```
Comment (rebel-jongho) [inline src/optimum/rbln/transformers/models/decoderonly/modeling_decoderonly.py:502]: how about using `@property` instead of multiple assigning??

File: src/optimum/rbln/transformers/models/decoderonly/modeling_decoderonly.py [hunk 8]
```diff
@@ -447,94 +449,66 @@ def prefill_forward(
...
+            self.block_tables = torch.arange(self.rbln_config.kvcache_num_blocks, dtype=torch.int16)
+        if self.use_local_attention:
+            self.local_block_tables = torch.tensor([0], dtype=torch.int16)
+        if self.rbln_config.use_attention_mask:
+            self.causal_mask = 1 - torch.triu(
+                torch.ones(1, 1, self.rbln_config.prefill_chunk_size, self.rbln_config.prefill_chunk_size), diagonal=1
             )
 
-        # NOTE(eunji): Use a decoder whose batch size matches the model's main batch size for compatibility.
-        self.decoder = self.decoders[self.rbln_config.batch_size]
```
Comment (rebel-jongho) [inline src/optimum/rbln/transformers/models/decoderonly/modeling_decoderonly.py:536]: @rebel-eunji Is it OK?

File: src/optimum/rbln/transformers/models/decoderonly/modeling_decoderonly.py [hunk 9]
```diff
@@ -658,205 +560,83 @@ def wrap_model_if_needed(cls, model: PreTrainedModel, rbln_config: "RBLNDecoderO
...
-        # Here we use meta tensor, for the memory efficiency.
-        meta_tensor_names = [name for name, _, _ in prefill_compile_config.input_info if "past_key_values" in name]
-        prefill_example_inputs = prefill_compile_config.get_dummy_inputs(fill=0, meta_tensor_names=meta_tensor_names)
-
         # Mark static tensors (self kv states)
         static_tensors = {}
-        for (name, _, _), tensor in zip(prefill_compile_config.input_info, prefill_example_inputs):
-            if "past_key_values" in name:
+        for (name, _, _), tensor in zip(compile_config.input_info, example_inputs):
+            if string_pattern in name:
```
Comment (rebel-jongho) [inline src/optimum/rbln/transformers/models/decoderonly/modeling_decoderonly.py:603]: Are there any models that don't use `past_key_values` as kv cache??

File: src/optimum/rbln/transformers/models/decoderonly/modeling_decoderonly.py (hunk 7 above)
Comment (rebel-thkim) [reply src/optimum/rbln/transformers/models/decoderonly/modeling_decoderonly.py:502]: I’m not sure, but I don’t know if this improves readability.
```
diff --git a/src/optimum/rbln/transformers/models/decoderonly/modeling_decoderonly.py b/src/optimum/rbln/transformers/models/decoderonly/modeling_decoderonly.py
index 47e5c61e..02201a70 100644
--- a/src/optimum/rbln/transformers/models/decoderonly/modeling_decoderonly.py
+++ b/src/optimum/rbln/transformers/models/decoderonly/modeling_decoderonly.py
@@ -17,6 +17,7 @@ from collections import deque
 from dataclasses import dataclass
 from pathlib import Path
 from typing import TYPE_CHECKING, Any, Callable, Deque, Dict, List, Optional, Tuple, Union
+from functools import cached_property
 
 import rebel
 import torch
@@ -498,8 +499,6 @@ class RBLNDecoderOnlyModel(RBLNModel, RBLNDecoderOnlyFlashAttentionMixin):
         self.prefill_decoder = RBLNPytorchRuntime(runtime=self.model[0])
 
         # attributes for prefill
-        self.use_global_attention = self.rbln_config.cache_impl in ["static", "hybrid"]
-        self.use_local_attention = self.rbln_config.cache_impl in ["sliding_window", "hybrid"]
         if self.use_global_attention:
             self.block_tables = torch.arange(self.rbln_config.kvcache_num_blocks, dtype=torch.int16)
         if self.use_local_attention:
@@ -509,6 +508,14 @@ class RBLNDecoderOnlyModel(RBLNModel, RBLNDecoderOnlyFlashAttentionMixin):
                 torch.ones(1, 1, self.rbln_config.prefill_chunk_size, self.rbln_config.prefill_chunk_size), diagonal=1
             )
 
+    @cached_property
+    def use_global_attention(self):
+        return self.rbln_config.cache_impl in ["static", "hybrid"]
+
+    @cached_property
+    def use_local_attention(self):
+        return self.rbln_config.cache_impl in ["sliding_window", "hybrid"]
+
```

File: src/optimum/rbln/transformers/models/decoderonly/modeling_decoderonly.py (hunk 9 above)
Comment (rebel-thkim) [reply src/optimum/rbln/transformers/models/decoderonly/modeling_decoderonly.py:603]: No, I think every models uses `past_key_values`.

Comment (rebel-thkim) [reply src/optimum/rbln/transformers/models/decoderonly/modeling_decoderonly.py:599]: Anyway I erase `phase` & `string_pattern` in args.

File: src/optimum/rbln/transformers/models/decoderonly/modeling_decoderonly.py [hunk 10]
```diff
@@ -998,8 +793,442 @@ def _update_rbln_config(
...
+
+        is_prefill = query_length > 1
+        # block_tables & local_block_tables
+        if rbln_config.use_global_attention:
+            idx = get_idx_from_name("block_tables")
+            max_block_cnt = rbln_config.max_seq_len // rbln_config.kvcache_block_size
+            input_info[idx] = ("block_tables", [max_block_cnt] if is_prefill else [batch_size, max_block_cnt], "int16")
+        if rbln_config.use_local_attention:
+            idx = get_idx_from_name("local_block_tables")
+            input_info[idx] = ("local_block_tables", [1] if is_prefill else [batch_size, 1], "int16")
```
Comment (rebel-jongho) [inline src/optimum/rbln/transformers/models/decoderonly/modeling_decoderonly.py:1201]: Just curious : why batch dim does not exist in input info, when prefill phase? Is it constraint of rebel-compiler?

I think if batch dim exists, these logics can easily integrated in DecoderOnly Model.

File: src/optimum/rbln/transformers/models/decoderonly/modeling_decoderonly.py [hunk 11]
```diff
@@ -998,8 +793,442 @@ def _update_rbln_config(
...
+
+        # local past_key_values shape
+        if rbln_config.use_local_attention:
+            local_kvcache_num_blocks = max(rbln_config.decoder_batch_sizes) if "decode" in rbln_config.phases else 1
+            num_layers = getattr(model_config, "num_hidden_layers") or getattr(model_config, "num_layers")
+            for i in range(num_layers * 2):
+                if rbln_config.sliding_window is not None and (i // 2) in rbln_config.sliding_window_layers:
+                    idx = i - num_layers * 2
+                    name, shape, dtype = input_info[idx]
+                    input_info[idx] = (name, [local_kvcache_num_blocks] + shape[1:], dtype)
```
Comment (rebel-jongho) [inline src/optimum/rbln/transformers/models/decoderonly/modeling_decoderonly.py:1224]: If these lines are just updating batch size, how about handle this in prefill phase(DecoderOnlyModel) as batch_size=1?

File: src/optimum/rbln/transformers/models/decoderonly/modeling_decoderonly.py (hunk 10 above)
Comment (rebel-thkim) [reply src/optimum/rbln/transformers/models/decoderonly/modeling_decoderonly.py:1201]: I agree, but the compiler team has asked me to remove the batch axis from the prefill.

**Comment by rebel-thkim** [comment]
airflow pr test done & pytest done

**Review by rebel-kblee — APPROVED** [review]
LGTM!!
